Shared/Extensions/URLExtensions.swift

```diff
@@ -338,62 +338,78 @@ extension URL {
 
 // Helpers to deal with ErrorPage URLs
 
-extension URL {
-    public static let errorPagePath = "errorpage"
-    public static let errorPageUrlParam = "url"
+public struct InternalURL {
+    public static let scheme = "internal"
+    public static let baseUrl = "\(scheme)://local"
+    public enum Path: String {
+        case errorPage = "errorpage"
+        case sessionRestore = "sessionrestore"
+        func matches(_ string: String) -> Bool { return string == self.rawValue }
+    }
+
+    public enum Param: String {
+        case url = "url"
+        func matches(_ string: String) -> Bool { return string == self.rawValue }
+    }
+
+    public let url: URL
+
+    public static func isValid(url: URL) -> Bool {
+        if AppConstants.IsRunningTest, url.path.contains("test-fixture/") {
+            return false
+        }
+
+        // TODO: (reader-mode-custom-scheme) remove this line when updating.
+        return url.absoluteString.hasPrefix("http://localhost:6571/") || InternalURL.scheme == url.scheme
+    }
+
+    public init?(_ url: URL) {
+        guard InternalURL.isValid(url: url) else {
+            return nil
+        }
 
-    public var isInternalErrorPage: Bool {
-        return isInternalScheme && path == "/\(URL.errorPagePath)"
+        self.url = url
+    }
+
+    public var isErrorPage: Bool {
+        return url.path == "/\(InternalURL.Path.errorPage)"
     }
 
     public var originalURLFromErrorPage: URL? {
-        let components = URLComponents(url: self, resolvingAgainstBaseURL: false)
-        if let queryURL = components?.queryItems?.find({ $0.name == URL.errorPageUrlParam })?.value {
+        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        if let queryURL = components?.queryItems?.find({ Param.url.matches($0.name) })?.value {
             return URL(string: queryURL)
         }
 
         return nil
     }
 
-    public var isInternalScheme: Bool {
```

Comment:
nit: I actually kinda preferred hanging this check off of `URL` versus having to invoke `InternalURL.isValid()`. I think my initial comment around this was to just shorten it to `.isInternal`.